### PR TITLE
[BEAM-5412][BEAM-5408] Fixes a bug that limited the size of TFRecords

### PR DESCRIPTION
**Pull request by chamikaramj**
Fixes a bug that limited the size of records read from TFRecord files that are GZIP compressed.

### CODE REVIEW

**Comment by chamikaramj** [comment]
R: @rangadi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java [hunk 1]
```diff
@@ -625,7 +625,19 @@ public int recordLength(byte[] data) {
       checkState(hashLong(length) == maskedCrc32OfLength, "Mismatch of length mask");
 
       ByteBuffer data = ByteBuffer.allocate((int) length);
-      checkState(inChannel.read(data) == length, "Invalid data");
+      long totalRead = 0;
```
Comment (rangadi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java]: You don't need totalRead(). Can just check 'data.remaining()'

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java [hunk 2]
```diff
@@ -625,7 +625,19 @@ public int recordLength(byte[] data) {
...
+      while (true) {
+        long read = inChannel.read(data);
+        if (read == 0) {
+          break;
+        }
+        totalRead += read;
```
Comment (rangadi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java]: remove.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TFRecordIOTest.java [hunk 3]
```diff
@@ -287,6 +288,18 @@ public void runTestRoundTripZlibFilesWithAuto() throws IOException {
     runTestRoundTrip(LARGE, 10, ".tfrecords", DEFLATE, AUTO);
   }
 
+  @Test
+  @Category(NeedsRunner.class)
+  public void runTestRoundTripLargeRecords() throws IOException {
```
Comment (rangadi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TFRecordIOTest.java]: Do both of the new tests fail or just the GZIP version without this fix? (just curious).

Comment (chamikaramj) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/TFRecordIOTest.java]: runTestRoundTripLargeRecordsGzip() fails without this change. Other test is for completeness.

**Comment by chamikaramj** [comment]
Breakage is unrelated: https://issues.apache.org/jira/browse/BEAM-5432

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java [hunk 4]
```diff
@@ -625,7 +625,19 @@ public int recordLength(byte[] data) {
       checkState(hashLong(length) == maskedCrc32OfLength, "Mismatch of length mask");
 
       ByteBuffer data = ByteBuffer.allocate((int) length);
-      checkState(inChannel.read(data) == length, "Invalid data");
+      long totalRead = 0;
+      while (true) {
+        long read = inChannel.read(data);
+        if (read == 0) {
```
[the remark this answers is not in the document]
Comment (rangadi) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java]: Right, that is true for blocking channels (which is usually the case).

**Comment by rangadi** [comment]
LGTM. Thanks @chamikaramj for the fix.

**Comment by chamikaramj** [comment]
Thanks.

**Comment by chamikaramj** [comment]
Retest this please

**Pull request merged**